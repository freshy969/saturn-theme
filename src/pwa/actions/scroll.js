import * as actionTypes from '../actionTypes';

export const windowHasScrolled = ({ direction }) => ({
  type: actionTypes.WINDOW_HAS_SCROLLED,
  direction
});

export const barsHaveHidden = () => ({
  type: actionTypes.BARS_HAVE_HIDDEN
});

export const barsHaveShown = () => ({
  type: actionTypes.BARS_HAVE_SHOWN
});
