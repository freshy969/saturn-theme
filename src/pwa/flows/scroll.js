import { throttle } from 'lodash';
import { getEnv } from 'mobx-state-tree';

let latestScroll = 0;

export const handleScroll = self =>
  throttle(async () => {
    const { latestDirection, setLatestDirection, showBar, hideBar } = self;

    const { fastdomPromised } = getEnv(self).theme;

    const { top } = await fastdomPromised.measure(() =>
      window.document.body.getBoundingClientRect(),
    );

    if (latestScroll === top && top !== 0) return;

    // Shows top bars if the scroll is too close to the top.
    if (top > -60) {
      showBar();
      if (latestDirection !== 'reset') setLatestDirection('reset');
    } else if (latestScroll > top) {
      // Shows/hiddes bars depending on scroll direction.
      if (latestDirection !== 'up') {
        hideBar();
        setLatestDirection('up');
      }
    } else if (latestDirection !== 'down') {
      if (latestDirection !== 'reset') showBar();
      setLatestDirection('down');
    }

    latestScroll = top;
  }, 100);

export const initializeScrollListener = async self => {
  window.addEventListener('scroll', handleScroll(self));
};

export const scrollMiddleware = (call, next) => {
  if (call.name === 'routeChangeSucceed') {
    call.tree.theme.scroll.setLatestDirection('reset');
  }

  next(call);
};
