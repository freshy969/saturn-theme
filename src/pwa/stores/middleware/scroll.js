export default (call, next) => {
  if (call.name === 'routeChangeSucceed') {
    call.tree.theme.scroll.setRouteChanged(true);
  }

  next(call);
};
