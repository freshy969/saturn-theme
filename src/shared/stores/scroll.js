import { types } from 'mobx-state-tree';

export default types
  .model('Scroll')
  .props({
    isBarHidden: false,
    latestScroll: 0,
    routeChanged: false,
  })
  .actions(self => ({
    hideBar() {
      self.isBarHidden = true;
    },
    showBar() {
      self.isBarHidden = false;
    },
    setRouteChanged(value) {
      self.routeChanged = value;
    },
    handleScroll(top) {
      if (self.latestScroll === top && top !== 0) return;

      if (top > -60) self.showBar();
      else if (self.routeChanged) {
        self.setRouteChanged(false);
        if (top < -60 && !self.isBarHidden) self.hideBar();
      } else if (top < self.latestScroll && !self.isBarHidden) self.hideBar();
      else if (top > self.latestScroll && self.isBarHidden) self.showBar();

      self.latestScroll = top;
    },
  }));
