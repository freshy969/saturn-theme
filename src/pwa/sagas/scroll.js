import { put, select, all, takeEvery } from 'redux-saga/effects';
import { throttle } from 'lodash';
import { dep } from 'worona-deps';
import { eventChannel } from 'redux-saga';
import * as actionTypes from '../actionTypes';
import * as actions from '../actions';

function* handleBarsOnScroll(action) {
  const { hiddenBars } = yield select(state => state.theme.scroll);
  const { direction } = action;

  if (direction === 'up' && !hiddenBars) {
    yield put(actions.scroll.barsHaveHidden());
  } else if (direction === 'down' && hiddenBars) {
    yield put(actions.scroll.barsHaveShown());
  }
}

const scroll = {
  latestDirection: null,
  latestScroll: 0
};

const windowScroll = () =>
  eventChannel(emitter => {
    const handleScroll = throttle(e => {
      const { top, bottom } = e.target.scrollingElement.getBoundingClientRect();
      const height = window.innerHeight;

      emitter({ top, bottom: bottom - height });
    }, 250);

    window.addEventListener('scroll', handleScroll);

    return () => window.removeEventListener('scroll', handleScroll);
  });

function* handleWindowScroll({ top, bottom }) {
  const { hiddenBars } = yield select(state => state.theme.scroll);
  const { latestDirection, latestScroll } = scroll;
  const isScrollingUp = latestScroll > top;
  scroll.latestScroll = top;

  // Shows top/bottom bars if the scroll is too close to the top/bottom.
  if (top > -60 || bottom < 700) {
    if (hiddenBars) yield put(actions.scroll.barsHaveShown());
  } else if (isScrollingUp) {
    // Shows/hiddes bars depending on scroll direction.
    if (latestDirection !== 'up') {
      yield put(actions.scroll.windowHasScrolled({ direction: 'up' }));
      scroll.latestDirection = 'up';
    }
  } else if (latestDirection !== 'down') {
    yield put(actions.scroll.windowHasScrolled({ direction: 'down' }));
    scroll.latestDirection = 'down';
  }
}

const handleRouteChange = () => {
  scroll.latestDirection = null;
  scroll.latestScroll = 0;
};

export default function* scrollSagas() {
  const windowScrollEvent = windowScroll();

  yield all([
    takeEvery(windowScrollEvent, handleWindowScroll),
    takeEvery(actionTypes.WINDOW_HAS_SCROLLED, handleBarsOnScroll),
    takeEvery(dep('connection', 'actionTypes', 'ROUTE_CHANGE_SUCCEED'), handleRouteChange)
  ]);
}
