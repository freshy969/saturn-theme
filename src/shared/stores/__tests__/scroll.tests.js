import scroll from '../scroll';

describe('Theme › Stores › Scroll', () => {
  test('hideBar changes the value of isBarHidden', () => {
    const self = scroll.create();

    expect(self.isBarHidden).toBe(false);
    self.hideBar();
    expect(self.isBarHidden).toBe(true);
  });

  test('showBar changes the value of isBarHidden', () => {
    const self = scroll.create({ isBarHidden: true });

    expect(self.isBarHidden).toBe(true);
    self.showBar();
    expect(self.isBarHidden).toBe(false);
  });

  test('setRouteChanged changes the value of routeChanged', () => {
    const self = scroll.create();

    expect(self.routeChanged).toBe(false);
    self.setRouteChanged(true);
    expect(self.routeChanged).toBe(true);
    self.setRouteChanged(false);
    expect(self.routeChanged).toBe(false);
  });

  test('setLatestScroll changes the value of latestScroll', () => {
    const self = scroll.create();

    expect(self.latestScroll).toBe(0);
    self.setLatestScroll(-42);
    expect(self.latestScroll).toBe(-42);
  });

  test("handleScroll doesn't do anything if `latestScroll === top`", () => {
    const self = scroll.create({ latestScroll: -42 });
    const showBar = jest.spyOn(self, 'showBar');
    const hideBar = jest.spyOn(self, 'hideBar');
    const setRouteChanged = jest.spyOn(self, 'setRouteChanged');
    const setLatestScroll = jest.spyOn(self, 'setLatestScroll');

    self.handleScroll(-42);
    expect(showBar).not.toHaveBeenCalled();
    expect(hideBar).not.toHaveBeenCalled();
    expect(setRouteChanged).not.toHaveBeenCalled();
    expect(setLatestScroll).not.toHaveBeenCalled();
  });

  test('If `latestScroll !== top` update the value of latestScroll', () => {
    const self = scroll.create();
    const setLatestScroll = jest.spyOn(self, 'setLatestScroll');

    self.handleScroll(-42);
    expect(setLatestScroll).toHaveBeenCalledWith(-42);
    self.handleScroll(0);
    expect(setLatestScroll).toHaveBeenCalledWith(0);
    self.handleScroll(-400);
    expect(setLatestScroll).toHaveBeenCalledWith(-400);
  });

  test('If the scroll is close to 0 and the bar is hidden call showBar', () => {
    const self = scroll.create({ isBarHidden: true });
    const showBar = jest.spyOn(self, 'showBar');

    self.handleScroll(-20);
    expect(showBar).toHaveBeenCalled();
  });

  test('If scrolling down and bar is not hidden call hideBar', () => {
    const self = scroll.create({ latestScroll: -60 });
    const hideBar = jest.spyOn(self, 'hideBar');

    self.handleScroll(-61);
    expect(hideBar).toHaveBeenCalled();
  });

  test('If scrolling up and bar is hidden call showBar', () => {
    const self = scroll.create({ isBarHidden: true, latestScroll: -400 });
    const showBar = jest.spyOn(self, 'showBar');

    self.handleScroll(-380);
    expect(showBar).toHaveBeenCalled();
  });

  test('After a route change call setRouteChanged', () => {
    const self = scroll.create({ isBarHidden: true, routeChanged: true });
    const setRouteChanged = jest.spyOn(self, 'setRouteChanged');

    self.handleScroll(-350);
    expect(setRouteChanged).toHaveBeenCalledWith(false);
    self.setRouteChanged(true);
    self.handleScroll(-20);
    expect(setRouteChanged).toHaveBeenCalledWith(false);
  });

  test('After a route change if scroll is close to 0 and bar is hidden call showBar', () => {
    const self = scroll.create({ isBarHidden: true, routeChanged: true });
    const showBar = jest.spyOn(self, 'showBar');

    self.handleScroll(-20);
    expect(showBar).toHaveBeenCalled();
  });

  test('After a route change if scrolled and bar is not hidden call hideBar', () => {
    const self = scroll.create({ routeChanged: true });
    const hideBar = jest.spyOn(self, 'hideBar');

    self.handleScroll(-200);
    expect(hideBar).toHaveBeenCalled();
  });

  test("After a route change if scrolled and bar is hidden don't call showBar or hideBar", () => {
    const self = scroll.create({ isBarHidden: true, routeChanged: true });
    const showBar = jest.spyOn(self, 'showBar');
    const hideBar = jest.spyOn(self, 'hideBar');
    const setRouteChanged = jest.spyOn(self, 'setRouteChanged');
    const setLatestScroll = jest.spyOn(self, 'setLatestScroll');

    self.handleScroll(-200);
    expect(hideBar).not.toHaveBeenCalled();
    expect(showBar).not.toHaveBeenCalled();
    expect(setRouteChanged).toHaveBeenCalledWith(false);
    expect(setLatestScroll).toHaveBeenCalledWith(-200);
  });
});
