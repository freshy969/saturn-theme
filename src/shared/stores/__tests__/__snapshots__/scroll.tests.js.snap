// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Theme › Shared › Stores › Scroll Initial snapshot has not changed 1`] = `
Object {
  "isBarHidden": false,
  "latestScroll": 0,
  "routeChanged": false,
}
`;
